### ISSUE TITLE
WFS provider does not iterate over all features

### ISSUE THREAD

**Issue by elpaso** [qgis/QGIS#39390]
See the test below, actual number of features is 143589, the server getfeature returns the first 1000 features but the noticeable thing is that if the test is run inside the QGIS console it still fails **unless the layer was previously added to the canvas**.

```python

    def testESRI(self):

        vl = QgsVectorLayer(" InvertAxisOrientation='1' restrictToRequestBBOX='1' srsname='EPSG:25833' typename='gewaessernetz_gewaessernetz_lfulg:Gewaessernetz' url='https://geoportal.umwelt.sachsen.de/arcgis/services/gewaessernetz/gewaessernetz_lfulg/MapServer/WFSServer' version='1.1.0'", 'layer', 'WFS')
        self.assertTrue(vl.isValid())

        self.assertTrue(vl.featureCount() > 1000)
       
        # This does help if run in the QGIS console:
        #QgsProject.instance().addMapLayers([vl])

        features = []
        for f in vl.getFeatures():
            features.append(f)

        self.assertEqual(len(features), vl.featureCount() )

        # This feature actually exists in the dataset but we cannot get it here
        exp = QgsExpression('"OBJECTID" = 1656082')
        request = QgsFeatureRequest(exp)
        feature = next(vl.getFeatures(request))

**Comment by rouault** [comment]
Enabling QgsProject.instance().addMapLayers([vl]) didn't change anything for me (which is reassuring), so I'm not sure if you actually tried exactly that.

But one key difference between using the unit test or adding the layer through QGIS UI, is that through QGIS UI, it will have version='auto' (and not 1.1.0), which will actually be evaluated as WFS 2.0, which does support paging. Whereas WFS 1.1 does not.

So to me it doesn't look like a bug

**Comment by elpaso** [comment]
> Enabling QgsProject.instance().addMapLayers([vl]) didn't change anything for me (which is reassuring), so I'm not sure if you actually tried exactly that.

No, I actually retrieved the layer from the project after loading it the "normal" way from the GUI.
Then I was able to iterate over it successfully from the QGIS Python console.

What I don't remember is if I opened the attribute table before doing that....


> 
> But one key difference between using the unit test or adding the layer through QGIS UI, is that through QGIS UI, it will have version='auto' (and not 1.1.0), which will actually be evaluated as WFS 2.0, which does support paging. Whereas WFS 1.1 does not.
> 
> So to me it doesn't look like a bug

mmm so you are saying that for WFS 1.1 we are not supposed to be able to iterate through all the features (because there is paging) from Python?  

What happens in the desktop? What trick is doing the desktop to make it work (besides the attr table loading)?

**Comment by rouault** [comment]
> No, I actually retrieved the layer from the project after loading it the "normal" way from the GUI.

hum, I've inserted your testESRI() in test_provider_wfs.py and enabling or disabling QgsProject.instance().addMapLayers([vl]) didn't change anything: only 1000 features retrieved

> mmm so you are saying that for WFS 1.1 we are not supposed to be able to iterate through all the features (because there is paging) from Python?

In QGIS UI, if you force WFS 1.1 in the WFS connection settings, you will only get 1000 features (if you pan the canvas, you can get 1000 features more due to the BBOX being different, but the working is a bit "random").
The WFS 1.1 spec doesn't support paging. A few servers implemented it as a sort of preview of WFS 2.0, but there's nothing in the GetCapabilities response of WFS 1.1 advertizing this, so currently the QGIS WFS provider doesn't attempt to do paging with WFS 1.1

> What happens in the desktop? What trick is doing the desktop to make it work (besides the attr table loading)?

I can't observe different behaviour in desktop vs unit test, if WFS version is set to the same value

**Comment by elpaso** [comment]
Ok, forget about it: I cannot reproduce it anymore and only god knows what I was doing.
